File: dev-run.exs
#!/usr/bin/env elixir

Code.require_file("dry_run.ex", __DIR__)

defmodule Run do
  use DryRun

  def main(args) do
    DryRun.init(args)

    if DryRun.enabled?() do
      IO.puts("DRY RUN - use --apply to execute")
    else
      IO.puts("EXECUTING")
      IO.puts("Requesting sudo access...")
      System.cmd("sudo", ["-v"])
    end

    runners_dir = "./runners"

    unless File.dir?(runners_dir) do
      IO.puts("Error: #{runners_dir} directory not found")
      System.halt(1)
    end

    runners = 
      File.ls!(runners_dir)
      |> Enum.sort()

    for runner <- runners do
      path = Path.join(runners_dir, runner)
      IO.puts("\n\e[35mRunning: #{runner}\e[0m")

      result = apply do
        System.shell("bash #{path}")
      end

      case result do
        :ok -> :ok

        {_, 0} ->
          IO.puts("\e[32m✓ #{runner} completed\e[0m")

        {_, exit_code} ->
          IO.puts("\e[31m✗ #{runner} failed with exit code #{exit_code}\e[0m")
          System.halt(exit_code)
      end
    end

    IO.puts("\n\e[32mAll runners completed\e[0m")
  end
end

Run.main(System.argv())

File: dev-run.pl
#!/usr/bin/env perl
use strict;
use warnings;

my $dry = !grep { $_ eq '--apply' } @ARGV;

if ($dry) {
    print "DRY RUN - use --apply to execute\n";
} else {
    print "EXECUTING\n";
    print "Requesting sudo access...\n";
    system("sudo -v") == 0 or die "Failed to get sudo access\n";

    print "Syncing package databases...\n";
    system("sudo pacman -Sy") == 0 or die "Failed to sync repos\n";

    # Keep sudo alive in background
    my $pid = fork();
    if ($pid == 0) {
        while (1) {
            sleep 60;
            system("sudo -n -v");
        }
    }

    # Kill background process on exit
    $SIG{INT} = $SIG{TERM} = sub { kill 9, $pid; exit; };
    END { kill 9, $pid if $pid; }
}

my $runners_dir = "./runners";

unless (-d $runners_dir) {
    die "Error: $runners_dir directory not found\n";
}

opendir(my $dh, $runners_dir) or die "Cannot open $runners_dir: $!";
my @runners = sort grep { !/^\./ && -f "$runners_dir/$_" } readdir($dh);
closedir($dh);

for my $runner (@runners) {
    my $path = "$runners_dir/$runner";
    print "\n\e[35mRunning: $runner\e[0m\n";

    if ($dry) {
        print "would execute: $path\n";
    } else {
        system($path);
        my $exit_code = $? >> 8;

        if ($exit_code == 0) {
            print "\e[32m✓ $runner completed\e[0m\n";
        } else {
            print "\e[31m✗ $runner failed with exit code $exit_code\e[0m\n";
            exit($exit_code);
        }
    }
}

print "\n\e[32mAll runners completed\e[0m\n";
